libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs

```diff
@@ -286,18 +297,47 @@ decodeCostModelFailHard lang = do
     Right cm -> pure cm
 {-# INLINEABLE decodeCostModelFailHard #-}
 
-getEvaluationContext :: CostModel -> PV1.EvaluationContext
+getEvaluationContext :: CostModel -> P.EvaluationContext
 getEvaluationContext (CostModel _ _ ec) = ec
 
 -- | See 'CostModels' for an explanation of how 'CostModelError' is used.
-newtype CostModelError = CostModelError PV1.CostModelApplyError
-  deriving stock (Show, Generic)
-
-instance Eq CostModelError where
-  CostModelError x1 == CostModelError x2 = show x1 == show x2
+newtype CostModelError = CostModelError P.CostModelApplyError
+  deriving stock (Eq, Show, Generic)
 
 instance Ord CostModelError where
-  CostModelError x1 <= CostModelError x2 = show x1 <= show x2
+  compare (CostModelError x1) (CostModelError x2) = comp x1 x2
+    where
+      comp (P.CMUnknownParamError err1) (P.CMUnknownParamError err2) = compare err1 err2
+      comp P.CMInternalReadError P.CMInternalReadError = EQ
+      comp (P.CMInternalWriteError err1) (P.CMInternalWriteError err2) = compare err1 err2
+      comp (P.CMTooFewParamsError e1 a1) (P.CMTooFewParamsError e2 a2) = compare e1 e2 <> compare a1 a2
+      comp cme1 cme2 = compare (tagOf cme1) (tagOf cme2)
+      tagOf :: P.CostModelApplyError -> Int
+      tagOf = \case
+        P.CMUnknownParamError {} -> 0
+        P.CMInternalReadError {} -> 1
+        P.CMInternalWriteError {} -> 2
+        P.CMTooFewParamsError {} -> 3
+
+instance EncCBOR CostModelError where
+  encCBOR =
+    encode . \case
+      CostModelError (P.CMUnknownParamError err) ->
+        Sum (CostModelError . P.CMUnknownParamError) 1 !> To err
```

Comment:
Why start at 1 and not at 0 ?

Reply:
Funny you say that. I wrote it initially starting with `0`, naturally. Then I looked at serialization of some predicate failure, which started at `1` and I thought to myself: "Oh, I guess we start at `1` for those decoders", so I switched it to start at `1` for "consistency". But now I can't even find that predicate failure :smile: 

I've switched to start at 0, thank you for bringing this up!